### PR TITLE
[transformers_modeling_backend] Enable the spmd_types sharding backend

### PR DESCRIPTION
**Pull request by mreso**
## What & why

  Wires the `spmd_types` sharding backend (`--parallelism.spmd_backend spmd_types`)
  into the HF transformers modeling backend, on top of #2679. `spmd_types` is the
  direction TorchTitan is migrating to (the DTensor `default`/`full_dtensor` paths
  are being removed). It also unblocks a config the DTensor backend cannot run:
  **DSA (GLM-5) + TP** — the indexer's `scatter_`/index ops have no DTensor eager
  dispatch (fail loud on `default`) but run natively on plain local tensors.

  ## How

  - **parallelize.py** — spmd_types mesh resolution (`dp_shard` axis), native
    vocab-parallel embedding, rope `no_typecheck` shield, module-protocol gate, and
    the CP k/v all-gather folded into the flex core kernels. `full_dtensor` is
    rejected (upstream is removing it).
  - **hf_sharding.py** — spmd redistribution sources; routes GQA/MLA (incl. CP) to
    typecheck-clean flex core kernels, DSA to its own core kernel.
  - **model.py** — `HFFlexAttnCoreKernel` / `HFMLAFlexAttnCoreKernel` /
    `HFDSAFlexAttnCoreKernel` run reshape / (split) / q-k-norm / RoPE + the flex HOP
    in a `local_map` region so the sharded reshape and RoPE are typecheck-clean.
    Handles per-head vs flat q/k norm (+ TP-aware flat RMSNorm for OLMoE),
    DeepSeek-V3 `rope_interleave`, the DSA dense mask + indexer, and Gemma4.
  - **trainer.py** — `global_vocab_size` for the vocab-parallel loss.

  Attention kernels remain HF's; only the sharding/typing wrapping is ours. No core
  TorchTitan files are modified.

  ## Verification

  Verified across every supported model (Qwen3, Mixtral, OLMoE, DeepSeek-V2-Lite/V3,
  GLM-4.7, GLM-5, Gemma4, dense, SFT) under FSDP / TP / EP / CP / TP+EP, both at
  runtime and with `--debug.spmd_typechecking`.

  **Numerical proof:** for each model, a 1-GPU deterministic fp32 run compares the
  full logit tensor between the `default` and `spmd_types` backends — **bit-identical
  (`max_abs_diff = 0.0`)** in every case. DSA + TP (impossible on `default`) trains
  cleanly on `spmd_types` (GLM-5 TP=2: 12.5 -> 6.8).

### CODE REVIEW

File: torchtitan/experiments/transformers_modeling_backend/hf_sharding.py [hunk 1]
```diff
@@ -68,6 +68,7 @@ def set_hf_sharding_configs(
     model: nn.Module,
     *,
     enable_sp: bool,
+    spmd_backend: str = "default",
```
Comment (fegin) [inline torchtitan/experiments/transformers_modeling_backend/hf_sharding.py]: I think we should just remove this flag and always use `spmd_types` unless there are concerns.
